### PR TITLE
Handle buffers based on their access type

### PR DESCRIPTION
**Pull request by mairooni**
#### Description

In this PR, each buffer that is created is matched with an access type (`READ_ONLY`, `WRITE_ONLY`, `READ_WRITE` or `NONE` if it cannot be deducted) depending on its usage. The access types are extracted from the `TornadoDataflowAnalysis` phase, which is applied during the sketching.  In the OpenCL and SPIR-V (OpenCL dispatcher) backends, the access type is passed as a flag during the buffer creation. Additionally, the free and used buffers are handled based on their access type. For example:  

```
Before this PR: 

free list: <bufferA, bufferB, bufferC>
used list: <bufferD>

In this PR (where bufferA and bufferB are READ_ONLY, bufferC is WRITE_ONLY and bufferD is READ_WRITE): 

free HashMap:   READ_ONLY -> <bufferA, bufferB>
                WRITE_ONLY -> <bufferC>
                READ_WRITE -> <>
                NONE-> <>
                           
used HashMap: READ_ONLY -> <>
              WRITE_ONLY -> <>
              READ_WRITE -> <bufferD>
              NONE-> <>                            

```
#### Backend/s tested

Mark the backends affected by this PR.

- [x] OpenCL
- [x] PTX
- [x] SPIRV

#### OS tested

Mark the OS where this PR is tested.

- [x] Linux
- [ ] OSx
- [ ] Windows

#### Did you check on FPGAs?

If it is applicable, check your changes on FPGAs.

- [ ] Yes
- [x] No

#### How to test the new patch?

Run `make-tests`

### CODE REVIEW

**Comment by jjfumero** [comment]
Thanks @mairooni. Any way to test this internally? Any debug flag?

File: tornado-drivers/drivers-common/src/main/java/uk/ac/manchester/tornado/drivers/common/TornadoBufferProvider.java [hunk 1]
```diff
@@ -44,8 +46,8 @@
 public abstract class TornadoBufferProvider {
 
     protected final TornadoDeviceContext deviceContext;
-    protected final List<BufferContainer> freeBuffers;
-    protected final List<BufferContainer> usedBuffers;
+    protected final HashMap<Access, ArrayList<BufferContainer>> freeBuffers;
```
Comment (jjfumero) [inline tornado-drivers/drivers-common/src/main/java/uk/ac/manchester/tornado/drivers/common/TornadoBufferProvider.java:49]: Use `Map` and allocate a ConcurrentHashMap. I think there is the possibility to run this code from multiple threads.

File: tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/OpenCL.java [hunk 2]
```diff
@@ -171,11 +171,11 @@ public static void run(Long executionContextId, OCLTornadoDevice tornadoDevice,
             switch (access) {
                 case READ_WRITE:
```
Comment (jjfumero) [inline tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/OpenCL.java:172]: When using `READ_WRITE` accessor, then the OpenCL flag should be `READ_WRITE` too.

**Review by jjfumero — CHANGES_REQUESTED** [review]

**Comment by jjfumero** [comment]
All test passing, including the GAIA multi-threaded use-case.

**Comment by mairooni** [comment]
@jjfumero your comments have been addressed. Please have a second look. Also, I included debug messages during the allocation and the buffer handling so that this PR is easier to debug.

File: tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/TornadoVM.java [hunk 3]
```diff
@@ -34,6 +34,7 @@
 import java.util.function.Consumer;
 
 import uk.ac.manchester.tornado.api.GridScheduler;
+import uk.ac.manchester.tornado.api.common.Access;
```
Comment (stratika) [inline tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/TornadoVM.java:37]:
```suggestion
```

Comment (stratika) [reply tornado-runtime/src/main/java/uk/ac/manchester/tornado/runtime/TornadoVM.java:37]: this is not used

File: tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/runtime/PTXTornadoDevice.java [hunk 4]
```diff
@@ -117,12 +117,12 @@ public TornadoSchedulingStrategy getPreferredSchedule() {
...
 
     @Override
-    public KernelStackFrame createKernelStackFrame(long executionPlanId, int numArgs) {
+    public KernelStackFrame createKernelStackFrame(long executionPlanId, int numArgs, Access access) {
         return getDeviceContext().getMemoryManager().createCallWrapper(executionPlanId, numArgs);
     }
 
     @Override
-    public XPUBuffer createOrReuseAtomicsBuffer(int[] arr) {
+    public XPUBuffer createOrReuseAtomicsBuffer(int[] arr, Access access) {
```
Comment (stratika) [inline tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/runtime/PTXTornadoDevice.java:125]: is the `access` argument used in any of the implementations of this method?

Comment (mairooni) [reply tornado-drivers/ptx/src/main/java/uk/ac/manchester/tornado/drivers/ptx/runtime/PTXTornadoDevice.java:125]: Yes, it is used in the implementation of the `OCLTornadoDevice` class

**Review by stratika — COMMENTED** [review]
From my side, it looks good. The only comment as far as I know that is pending is regarding the usage of ConcurrentHashMap. Any update on that front?

**Comment by jjfumero** [comment]
We actually don't need the concurrent hash map. So I am good with this PR.

**Review by stratika — APPROVED** [review]
Perfect, thanks @mairooni. Ready to merge.

**Review by jjfumero — APPROVED** [review]

**Pull request merged**